Add CssSelectorBridge

ORelio opened the pull request:
Hi rss-bridge team 🙂

In this PR I'm presenting some kind of utility bridge based on an idea I got while working on WordPress bridge.
CustomBridge is an even more generic approach than WordPress bridge's custom content selector feature (#3125).
With this bridge and only 2 HTML selectors, you can quickly generate a feed from blogs lacking a RSS feed:

* URL selector to select recent articles on home page
* Content selector to select article content on article page

Of course it is only useful to advanced users who know how to make content selectors. I've been using it for like 6 months to quickly generate feeds for small blogs I'd be otherwise too lazy to make a bridge for. I would not recommend using it for anything but small blogs that miss a RSS feed or voluntarily disabled it (sigh).

![image](https://github.com/RSS-Bridge/rss-bridge/assets/1128206/5376f096-07c4-43ca-8b3d-3d1d0e3438e6)

Under the hood, the bridge is designed to be pretty simple and generic.
As such, only minimal content is included in the feed:

* Item title is automatically taken from page `<title>`, which is usually correct because it's vital for SEO.
* Item date is missing because of too many potential formats; but most RSS aggregators will fill the gap.

There are optional parameters to exclude blog title from article title, and unwanted items from article content.
No other features are planned, as they would be too specific, and this bridge is not intended to replace tailor-made bridges.

I'm sharing this bridge in the hope that it will be useful to other advanced users 😉

ORelio committed:
Add CustomBridge

For advanced users. Create RSS feed using HTML selectors.

dvikan commented:
This is a good idea. Reminds me of those other rss generators where you insert css selectors or "placeholders".

The concept "Custom bridge" is already used by rssbridge. Maybe find another name. perhaps `SelectorBridge` or `CssSelectorBridge` 

Also perhaps not fetch each item but instead just produce a list of new items. Maybe thats just me but im mostly intererested in being alerted for new items.

ORelio commented:
> The concept "Custom bridge" is already used by rssbridge. Maybe find another name. perhaps SelectorBridge or CssSelectorBridge

OK. Will do.

> Also perhaps not fetch each item but instead just produce a list of new items.

Why not. I see a potential issue with listing items without fetching content: title is taken from each page's `<title>` attribute. I'll test against the sites I'm using the bridge with, to see if taking title from hyperlinks instead is reliable enough for a basic truncated feed.

I'll update the pull request once I've done enough testing.

dvikan commented:
i am speculating that the most common use case is to scrape e.g. a frontpage which has a list of items (title, url and possibly date).

there is extra work required to scrape the item content too. but it's up to you

dvikan commented:
I put a copy of your bridge here: https://rss-bridge.org/bridge01/#bridge-CssSelectorTestBridge (for testing)

See also https://rsseverything.com/newfeed

See also https://politepol.com/en/

ORelio committed:
[CssSelectorBridge] Refactor, Allow Unexpanded

Rename bridge to CssSelectorBridge
Allow unexpanded feed, i.e. make feed from home page only (1 request)
Refactor bridge to put most of the code into protected functions
Makes the code more maintainable and allows inheritance for variants

ORelio committed:
[CssSelectorBridge] Fix linting

ORelio commented:
Using text inside the `<a>` as title for each element gave satisfying results on the sites I tried, although that title may be truncated. So I made the content selector optional, and if not specified, articles are not expanded.

I also took that opportunity to reorganize the code in functions. Besides making it more readable/maintainable, I plan to make some variants of the bridge that would inherit from the `CssSelectorBridge`.

An example of planned variant: I encountered some sites that were loading their article list dynamically from an API. I wondered how Google could index that, since it doesn't evaluate scripts. Turns out these sites have a `sitemap.xml` that we could use the same way Google does 😏. That variant would give us a `SitemapBridge` inheriting from `CssSelectorBridge`... but that will be for another pull request 😄

dvikan commented:
also relevant: https://feed-me-up-scotty.vincenttunru.com/

dvikan commented:
any progress here? this bridge is gonna be really cool and useful

ORelio commented:
Well, I did implement the suggested changes.
Anything else you would need? :slightly_smiling_face:

dvikan commented:
alright cool cool.

i deployed it to https://rss-bridge.org/bridge01/#bridge-CssSelectorBridge if anyone wants to test it out

Strikelab commented:
Really nice bridge !

ORelio commented:
Glad to see it's useful not just to me 😄

Strikelab commented:
> Glad to see it's useful not just to me 😄

Really usefull to extract documentation from pages like developer.mozilla.org and store them in FreshRss .
Need several trys to get the good output, but that's definitively a really nice feature !
Thanks !

ghost commented:
@ORelio Is the bridge supposed to extract several feed items (i.e. articles) by default? I tested and can only get the first item:

```
Site URL: Home page with latest articles: https://www.eurogamer.net/topics/digital-foundry
Selector for article links or their parent elements: div.page_content
```

I'd have expected all 25 post elements (`.link_overlay`). In general I only want the webpage title and article URLs in a feed.

ORelio commented:
@07416 There's a `limit` parameter that defaults to 10 when not set. Make sure it's not set to 1. The bridge takes all matching items within the limit.

https://rss-bridge.org/bridge01/?action=display&bridge=CssSelectorBridge&home_page=https%3A%2F%2Fwww.eurogamer.net%2Ftopics%2Fdigital-foundry&url_selector=.link_overlay&content_selector=article&content_cleanup=script,%20div.breadcrumbs,%20div.headline_details,%20div.metadata,%20div.mypop-header-wrapper,%20div.social_share&title_cleanup=+|+Eurogamer.net&format=Html

Seems to work fine.

ghost commented:
@ORelio Thanks, I only changed the selector to `.link_overlay`. 

(if the `limit defaults to 10, the RSS-bridge page should should show the value as a "grayed out" placeholder text in the field)

Your rule is much more complex — at the time I mostly don't have need for extracting full article content, I view relevant pages in browser (trivial in `newsboat`).

---

I hope that the bridge will take over as the de facto feed-generator — there doesn't appear to be many viable self-hostable solutions which allow comprehensive filtering.

ORelio commented:
> at the time I mostly don't have need for extracting full article content

If you only want links, it's also possible:

https://rss-bridge.org/bridge01/?action=display&bridge=CssSelectorBridge&home_page=https%3A%2F%2Fwww.eurogamer.net%2Ftopics%2Fdigital-foundry&url_selector=a.title_link&format=Html

Regarding limit:
> (if the `limit defaults to 10, the RSS-bridge page should should show the value as a "grayed out" placeholder text in the field)

On bridge card, limit is an automatically generated field.
```php
   const PARAMETERS = [
        [
            'home_page' => [
                'name' => 'Site URL: Home page with latest articles',
                'exampleValue' => 'https://example.com/blog/',
                'required' => true
            ],
//[...]
            'limit' => self::LIMIT
        ]
    ];
```
But the default value is up to each bridge, in mine I use 10 when limit is not set.
```php
$limit = $this->getInput('limit') ?? 10;
```
So by the time the limit field is shown on bridge card, RSS-Bridge does not know yet what value to show as placeholder.
